score.go

```diff
@@ -111,54 +104,82 @@ func (d *indexData) scoreFile(fileMatch *FileMatch, doc uint32, mt matchTree, kn
 	addScore("repo-rank", scoreRepoRankFactor*float64(md.Rank)/maxUInt16)
 
 	if opts.DebugScore {
-		fileMatch.Debug = strings.TrimSuffix(fileMatch.Debug, ", ")
+		fileMatch.Debug = fmt.Sprintf("score: %.2f <- %s", fileMatch.Score, strings.TrimSuffix(fileMatch.Debug, ", "))
 	}
 }
 
-// scoreFileUsingBM25 computes a score for the file match using an approximation to BM25, the most common scoring
-// algorithm for text search: https://en.wikipedia.org/wiki/Okapi_BM25. It implements all parts of the formula
-// except inverse document frequency (idf), since we don't have access to global term frequency statistics.
+// calculateTermFrequency computes the term frequency for the file match.
 //
-// Filename matches count twice as much as content matches. This mimics a common text search strategy where you
-// 'boost' matches on document titles.
-//
-// This scoring strategy ignores all other signals including document ranks. This keeps things simple for now,
-// since BM25 is not normalized and can be tricky to combine with other scoring signals.
-func (d *indexData) scoreFileUsingBM25(fileMatch *FileMatch, doc uint32, cands []*candidateMatch, opts *SearchOptions) {
+// Filename matches count more than content matches. This mimics a common text
+// search strategy where you 'boost' matches on document titles.
+func calculateTermFrequency(fileMatch *FileMatch, cands []*candidateMatch, df termDocumentFrequency) {
 	// Treat each candidate match as a term and compute the frequencies. For now, ignore case
 	// sensitivity and treat filenames and symbols the same as content.
 	termFreqs := map[string]int{}
 	for _, cand := range cands {
 		term := string(cand.substrLowered)
 
+		if _, ok := termFreqs[term]; !ok {
+			df[term] += 1
```

Comment:
I just noticed -- shouldn't we only be increasing `df` once per file, instead of once per candidate match? We consider a file to be a "document" for the BM25 calculation, not a chunk match.

Reply:
The check `if _, ok := termFreqs[term]; !ok {` should make sure we only increase it once per document. We can pull that out of the loop to make it more obvious at the cost of an additional loop. However, there shouldn't be too many unique terms so it's probably fine. I also added a new unit case.